src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -131,6 +162,128 @@ public class FhirCosmosSearchService : SearchService
+            {
+                throw new SearchOperationNotSupportedException(Resources.ChainedExpressionNotSupported);
+            }
+
+            var chainedReferences = new List<Expression>();
+            SearchOptions chainedOptions = searchOptions.Clone();
+            chainedOptions.CountOnly = false;
+            chainedOptions.MaxItemCount = _chainedSearchMaxSubqueryItemLimit;
+            chainedOptions.MaxItemCountSpecifiedByClient = false;
+
+            // Local function for recursion, this lets us walk to the end of the expression to start filtering,
+            // the results are used to filter against the parent layer.
+            async Task<Expression> RecursiveLookup(ChainedExpression expression)
+            {
+                Expression criteria = expression.Expression;
+
+                if (expression.Expression is ChainedExpression innerChained)
+                {
+                    criteria = await RecursiveLookup(innerChained);
+                }
+
+                if (criteria == null)
+                {
+                    // No items where returned by the sub-queries, break
+                    return null;
+                }
+
+                string filteredType = expression.TargetResourceTypes.FirstOrDefault();
+                var includeExpressions = new List<IncludeExpression>();
+
+                if (expression.Reversed)
+                {
+                    // When reversed we'll use the Include expression code to return the ids
+                    // in the search index on the matched resources
+                    foreach (var targetInclude in expression.TargetResourceTypes)
+                    {
+                        includeExpressions.Add(Expression.Include(
+                            expression.ResourceTypes,
+                            expression.ReferenceSearchParameter,
+                            null,
+                            targetInclude,
+                            expression.TargetResourceTypes,
+                            false,
+                            false,
+                            false));
+                    }
+
+                    // When reversed the ids from the sub-query will match the base resource type
+                    filteredType = expression.ResourceTypes.FirstOrDefault();
+                }
+
+                MultiaryExpression filterExpression = Expression.And(
+                    Expression.SearchParameter(_resourceTypeSearchParameter, Expression.StringEquals(FieldName.TokenCode, null, filteredType, false)),
+                    criteria);
+
+                chainedOptions.Expression = filterExpression;
+
+                var chainedResults = await ExecuteSearchAsync<FhirCosmosResourceWrapper>(
+                    _queryBuilder.BuildSqlQuerySpec(chainedOptions, includeExpressions, idsOnly: true),
```

Comment:
>idsOnly: true [](start = 88, length = 13)

It looks like you use ids only if `!expression.Reversed` so no point to fetch them, so maybe this can be enum: `Default`, `Id`, `Nothing`(IncludeOnly)?